src/libasr/pass/replace_symbolic.cpp

```diff
@@ -294,405 +291,196 @@ class ReplaceSymbolicVisitor : public PassUtils::PassVisitor<ReplaceSymbolicVisi
-                            ASR::IntrinsicFunction_t* s = ASR::down_cast<ASR::IntrinsicFunction_t>(value);
-                            this->visit_IntrinsicFunction(*s);
-                            ASR::symbol_t* var_sym = current_scope->get_symbol(symengine_stack.pop());
-                            value = ASRUtils::EXPR(ASR::make_Var_t(al, x.base.base.loc, var_sym));
-                        } else if (ASR::is_a<ASR::Cast_t>(*value)) {
-                            ASR::Cast_t* s = ASR::down_cast<ASR::Cast_t>(value);
-                            this->visit_Cast(*s);
-                            ASR::symbol_t* var_sym = current_scope->get_symbol(symengine_stack.pop());
-                            value = ASRUtils::EXPR(ASR::make_Var_t(al, x.base.base.loc, var_sym));
-                        }
-                        perform_symbolic_unary_operation(al, x.base.base.loc, module_scope, "basic_abs",
-                            x.m_target, value);
-                        break;
-                    }
-                    case LCompilers::ASRUtils::IntrinsicFunctions::SymbolicExpand: {
-                        ASR::expr_t* value = intrinsic_func->m_args[0];
-
-                        if (ASR::is_a<ASR::IntrinsicFunction_t>(*value)) {
-                            ASR::IntrinsicFunction_t* s = ASR::down_cast<ASR::IntrinsicFunction_t>(value);
-                            this->visit_IntrinsicFunction(*s);
-                            ASR::symbol_t* var_sym = current_scope->get_symbol(symengine_stack.pop());
-                            value = ASRUtils::EXPR(ASR::make_Var_t(al, x.base.base.loc, var_sym));
-                        } else if (ASR::is_a<ASR::Cast_t>(*value)) {
-                            ASR::Cast_t* s = ASR::down_cast<ASR::Cast_t>(value);
-                            this->visit_Cast(*s);
-                            ASR::symbol_t* var_sym = current_scope->get_symbol(symengine_stack.pop());
-                            value = ASRUtils::EXPR(ASR::make_Var_t(al, x.base.base.loc, var_sym));
-                        }
-                        perform_symbolic_unary_operation(al, x.base.base.loc, module_scope, "basic_expand",
-                            x.m_target, value);
-                        break;
-                    }
-                    default: {
-                        throw LCompilersException("IntrinsicFunction: `"
-                            + ASRUtils::get_intrinsic_name(intrinsic_id)
-                            + "` is not implemented");
-                    }
-                }
+    void visit_Assignment(const ASR::Assignment_t &x) {
+        SymbolTable* module_scope = current_scope->parent;
+        if (ASR::is_a<ASR::IntrinsicFunction_t>(*x.m_value)) {
+            ASR::IntrinsicFunction_t* intrinsic_func = ASR::down_cast<ASR::IntrinsicFunction_t>(x.m_value);
+            if (intrinsic_func->m_type->type == ASR::ttypeType::SymbolicExpression) {
```

Comment:
Yup, I was talking about this in first review. As you did, we have to only allow the `SymbolicExpression` (which we later handle to CPtr). Otherwise, we might get Type mismatch for the BindC function (which might be difficult to debug).

Reply:
Yes this has been taken care of in all `visit_Assignment` (line 483) , visit_Print (line 626) and visit_IntrinsicFunction (line 711) functions through a type check for SymbolicExpression.